addons/ai/functions/fnc_garrison.sqf

```diff
@@ -0,0 +1,173 @@
+/*
+ * Author: alganthe
+ * Garrison function used to garrison AI inside buildings.
+ *
+ * Arguments:
+ * 0: The building(s) nearest this position are used <POSITION>
+ * 1: Limit the building search to those type of building <ARRAY>
+ * 2: Units that will be garrisoned <ARRAY>
+ * 3: Radius to fill building(s) <SCALAR> (default: 50)
+ * 4: 0: even filling, 1: building by building, 2: random filling <SCALAR> (default: 0)
+ * 5: True to fill building(s) from top to bottom <BOOL> (default: false)
+
+ * Return Value:
+ * Units not garrisoned <ARRAY>
+ *
+ * Example:
+ * [position, nil, [unit1, unit2, unit3, unitN], 200, 1, false] call ace_ai_fnc_garrison
+ *
+ * Public: Yes
+*/
+#include "script_component.hpp"
+
+params [["_startingPos",[0,0,0], [[]], 3], ["_buildingTypes", ["Building"], [[]]], ["_unitsArray", [], [[]]], ["_fillingRadius", 50, [0]], ["_fillingType", 0, [0]], ["_topDownFilling", false, [true]]];
+
+_unitsArray = _unitsArray select {alive _x && {!isPlayer _x}};
+
+if (_startingPos isEqualTo [0,0,0]) exitWith {
+    [LSTRING(GarrisonInvalidPosition)] call EFUNC(common,displayTextStructured);
+};
+
+if (count _unitsArray == 0 || {isNull (_unitsArray select 0)}) exitWith {
+    [LSTRING(GarrisonNoUnits)] call EFUNC(common,displayTextStructured);
+};
+
+private _buildings = [];
+
+if (_fillingRadius < 50) then {
+    _buildings = nearestObjects [_startingPos, _buildingTypes, 50];
+} else {
+    _buildings = nearestObjects [_startingPos, _buildingTypes, _fillingRadius];
+    _buildings = _buildings call BIS_fnc_arrayShuffle;
+};
```

Comment:
This whole thing would be neater as:
```sqf
private _buildings = nearestObjects [_startingPos, _buildingTypes, _fillingRadius];
if (_fillingRadius >= 50) then {
    _buildings = _buildings call BIS_fnc_arrayShuffle;
};
```

There is also [`CBA_fnc_shuffle`](https://github.com/CBATeam/CBA_A3/blob/master/addons/arrays/fnc_shuffle.sqf) that is probably faster than BI's function.